### PR TITLE
Updates to FATES zenith angles on the first step of non-continue restarts

### PR DESCRIPTION
**Pull request by rgknox**
These changes ensure that zenith angles are passed to FATES on non-continue restarts on the first step. The legacy FATES radiative transfer scheme (Norman) was not sensitive to this, but the newer two-stream scheme needs this information at this point in the call sequence.

[non-BFB] for FATES tests.

### CODE REVIEW

File: components/elm/src/main/elm_driver.F90 [hunk 1]
```diff
@@ -1365,7 +1369,20 @@ subroutine elm_drv(doalb, nextsw_cday, declinp1, declin, rstwr, nlend, rdate)
...
        ! Determine albedos for next time step
        ! ============================================================================
+       
+       if (use_fates .and. .not.doalb ) then
+
+          is_cold_start =  (nsrest == nsrStartup .and. finidat == ' ')
+          
+          if ( (is_cold_start .and. get_nstep() == 1) .or. & 
+               ((fates_radiation_model == 'twostream') .and. (get_nstep()== 1) .and. (.not.use_fates_sp) &
+               .and. (.not.is_cold_start) .and. (nsrest == nsrStartup)) ) then
```
Comment (bishtgautam) [inline components/elm/src/main/elm_driver.F90]: @rgknox , can we simplify these conditionals or break them down so they are a bit more manageable to understand?

```fortran
if (get_nstep() == 1) then

   do_update = .false.

   if (is_cold_start)
       do_update = .true.
   else if (<check-second-conditional>)
       do_update = .true.
   endif

   if (do_update) then
      call alm_fates%wrap_canopy_radiation(bounds_clump,surfalb_vars,nextsw_cday,declinp1)
   endif

endif 
```

Comment (rgknox) [reply components/elm/src/main/elm_driver.F90]: yeah, good idea! I'll submit a change

**Comment by rgknox** [comment]
Hehe, @bishtgautam , I was able to make it way simpler.  I don't know how all that extra logic got in there.  Anyway, take a look and see what you think.  Thanks for the suggestion

**Comment by rgknox** [comment]
NorESM has these land-model tests called Land Tuning Mode.   I think that these tests simulate frequency of the radiation boundary conditions when running coupled with the atmosphere, although it still uses a data atmosphere for the information.  Do we have these types of tests in E3SM? I would add them to the FATES test list.

**Comment by bishtgautam** [comment]
Hi @rgknox, thanks for simplifying the logic. How about also adding a comment explaining what is happening? something like the following, based on your [first comment](https://github.com/E3SM-Project/E3SM/pull/7899#issue-3648945772).

```
! On the first step, send the solar zenith angles to FATES on non-continue restarts
! for the two-stream radiation scheme.
```

**Comment by glemieux** [comment]
@bishtgautam I talked with Ryan and I'll add your comment and then kick off e3sm land developer tests soon.

**Comment by glemieux** [comment]
Regression testing `e3sm_land_developer` against the `master` baseline is underway on perlmutter.

**Comment by glemieux** [comment]
Regression testing against the `master` baseline on perlmutter is complete and showing only B4B differences for the subset of FATES test mods.  @rgknox is this expected? 

Results: `/global/homes/g/glemieux/scratch/e3sm-tests/pr7899-zenith.fates.pm-cpu..E3d3963866b-F5bb36cba2`

**Comment by glemieux** [comment]
> Regression testing against the `master` baseline on perlmutter is complete and showing only B4B differences for the subset of FATES test mods. @rgknox is this expected?
> 
> Results: `/global/homes/g/glemieux/scratch/e3sm-tests/pr7899-zenith.fates.pm-cpu..E3d3963866b-F5bb36cba2`

I'm thinking this is likely due to the fact that I didn't rebase this against `master` prior to testing.  The code only impacts two-stream runs and Norman is currently the default.  I'm currently retesting this after rebasing.

**Comment by glemieux** [comment]
> I'm thinking this is likely due to the fact that I didn't rebase this against `master` prior to testing. The code only impacts two-stream runs and Norman is currently the default. I'm currently retesting this after rebasing.

I'm still seeing unexpected diffs.  I'm generating my own baseline using `master` to triple check, just in case the `master`  baseline wasn't built with the commits I think it's should be representing.

File: components/elm/src/main/elm_driver.F90 [hunk 2]
```diff
@@ -1366,7 +1369,16 @@ subroutine elm_drv(doalb, nextsw_cday, declinp1, declin, rstwr, nlend, rdate)
...
        ! ============================================================================
-
+       
+       ! On the first step, send the solar zenith angles to FATES on non-continue restarts
+       ! for the two-stream radiation scheme.
+       if (use_fates .and. .not.doalb .and. get_nstep() == 1 .and. nsrest == nsrStartup) then
+          if ( (trim(finidat) == '') .or. & 
+               (fates_radiation_model=='twostream' .and. .not.use_fates_sp )) then
+             call alm_fates%wrap_canopy_radiation(bounds_clump,surfalb_vars,nextsw_cday,declinp1)
+          end if
```
Comment (glemieux) [inline components/elm/src/main/elm_driver.F90]: Reviewing the FATES test `lnd_in` files, the default for `finidat` is always empty, so this check will always evaluate `TRUE` for FATES testmods.  As such, I believe we need to nest these checks (i.e. remove the `.or.`) since we want to fail if we are not running twostream or are running sp mode: 
```suggestion
          if (trim(finidat) == '') then
             if ( fates_radiation_model=='twostream' .and. .not.use_fates_sp ) then
                call alm_fates%wrap_canopy_radiation(bounds_clump,surfalb_vars,nextsw_cday,declinp1)
             end if
          end if
```

**Review by glemieux — CHANGES_REQUESTED** [review]

File: components/elm/src/main/elm_driver.F90 (hunk 2 above)
Comment (glemieux) [reply components/elm/src/main/elm_driver.F90]: @rgknox with the above change, the `e3sm_land_developer` regression tests against my master baseline are B4B.  Note that we don't have any two stream tests for in the `e3sm_land_developer` suite yet.

Comment (glemieux) [reply components/elm/src/main/elm_driver.F90]: Testing this against the fates suite of tests shows only DIFFs in the `twostream` test mod as expected.

Comment (rgknox) [reply components/elm/src/main/elm_driver.F90]: @glemieux my take is that this set of changes was important for situations when finitdat /= '', hence the "or".  The tests may not generate this condition, but it's still meaningful. Also, per our discussion, I don't think we need to filter out SP runs. We only filtered out SP runs for CTSM to preserve b4b behavior due to ongoing calibration efforts.

**Review by glemieux — APPROVED** [review]

**Comment by glemieux** [comment]
@peterdschwartz this is good to move forward.

**Comment by peterdschwartz** [comment]
what are the expected diffs?  Running on next on chrysalis gives:
```
  ERS_D_Ld15.f45_g37.IELMFATES.chrysalis_intel.elm-fates_cold_treedamage (Overall: DIFF) details:
  ERS_Ld20.f45_f45.IELMFATES.chrysalis_intel.elm-fates (Overall: DIFF) details:
  ERS_Ld30.f45_f45.IELMFATES.chrysalis_intel.elm-fates_satphen (Overall: DIFF) details:
  ERS_Ld30.f45_g37.IELMFATES.chrysalis_intel.elm-fates_cold_sizeagemort (Overall: DIFF) details:
  SMS_D_Ld20.f45_f45.IELMFATES.chrysalis_intel.elm-fates_rd (Overall: DIFF) details:
  SMS_Ld20.f45_f45.IELMFATES.chrysalis_intel.elm-fates_eca (Overall: DIFF) details:
  SMS_Ld5_PS.f19_g16.IELMFATES.chrysalis_intel.elm-fates_cold (Overall: DIFF) details:
```

**Comment by glemieux** [comment]
@peterdschwartz yes, these DIFFs are expected.  Thanks for checking.

**Comment by peterdschwartz** [comment]
Some machine issues last night/today. will merge to next tomorrow if resolved

**Comment by peterdschwartz** [comment]
merged to next

**Comment by peterdschwartz** [comment]
merged to master
